### ISSUE TITLE
REF: Pearson estimate of scale parameter/Tweedie parameter with weights

### ISSUE THREAD

**Issue by thequackdaddy** [statsmodels/statsmodels#2915]
I'm a little confused over how to interpret the Pearson estimates for the scale parameter and a shape parameter in a frequency dataset. 

The Pearson estimate of scale is as follows:

![formula][1]

  [1]: https://chart.googleapis.com/chart?cht=tx&chl=%5Cphi_%7Bpearson%7D%20%3D%20%5Cfrac%7B%5Csum%7B%5Cfrac%7Bw_i%20(Y_i%20-%20%5Cmu_i)%5E2%7D%7BV(%5Cmu_i)%7D%7D%7D%7Bdf_%7Bresid%7D%7D

If I have 1 record representing 1 year of a policy and it has 1 claim, then the frequency (which is the same as endog and Y) equals 1.

If I split that record into 2 separate 6-month records, then I now have 1 record with a frequency of 2 and a separate record with a frequency of 0. (The claim only occurs once... the only difference is how the data is laid out). Mu is the same between all the scenarios. It becomes pretty intuitive to see that you will get a different estimate of phi based off of my (somewhat) arbitrary splits in the policy. 

Is there an adjustment that we could do to the Pearson estimate so that 1 12-month record and 2 six-month records, with 1 annual claim in each scenario, would give the same estimate of scale?

**Comment by josef-pkt** [comment]
Do you mean the rate (case per period) with `frequency`?  (frequency makes me think of time series)

I haven't done the math yet, so I'm thinking mainly of Poisson and WLS for the following.

The scale parameter phi is relative to something. In general rescaling the process changes the `scale`. 
In terms of exposure (e.g. total time for a Poisson process)
In the first case exposure is 1 (year). In the second case exposure is half a year, so units will be per half year, i.e. the arrival rate = 2 per half year for the first observation. 
(... explanation doesn't work because Poisson process doesn't have exess dispersion, so I'm thinking more about mean than dispersion.)

When we predict the mean then we want to standardize the exposure to say for example what's the arrival/claim rate per year. (Like quoting annualized interest rate, even if it's calculated monthly.)

There should be a similar "annualization" for the dispersion or variance but I haven't figured out completely how aggregation works if there is excess dispersion. (QMLE instead of MLE)

For variance (not dispersion) we have that the variance of the sum is n times the variance of each splitted fragment. (back of the envelope calculation)
but in Tweedie or GLM we have the separation between variance function (as function of mean) and the dispersion phi.

another problem: only Poisson distribution is the count distribution that is time homogeneous, other count distributions don't split up into subperiods in the simple additive way.

however, Tweedie as compound Poisson is a time homogeneous process. so we need to link the dispersion coefficient of the tweedie distribution to the sum of claim sizes per standardized unit of time. 

based on another quick derivation by hand: the dispersion per subperiod would have to be divided by n**(p-1) to arrive at the dispersion per period, if there are n subperiods per period.

something like this, I haven't seen it in any reference.  (thinking while writing, no guarantees)

We need to check at least the distribution of the sum for some special cases like Gamma.

**Comment by thequackdaddy** [comment]
@josef-pkt As usual, you've hit the nail on the head with your interpretation of what I'm writing. By frequency, I mean the number of claims reported per year of policy in-force. (Actuarial texts generally define frequency as claims per unit of exposure... so much different usage for "frequency" and "exposure" than a statistician--or economist--would like.) 

If you stumble across any papers on annualization of phi/scale parameter (a quick google search wasn't fruitful), let me know. Maybe "annualization" isn't the right word...

**Comment by josef-pkt** [comment]
I didn't find anything yet to verify the dispersion calculation. However, the properties of the variance when aggregating are standard. for example the variance of a compound poisson process is proportional to the time length. So the sub periods should have 1/n times the variance of the full period sum. 
What this implies for the dispersion coefficient depends on the shape of the variance function, since the mean also increases also linearly in aggregating subperiods

**Comment by josef-pkt** [comment]
phi = lambda**(1-p) / (2 - p)   
where   p = (alpha + 2) / (alpha + 1)  alpha is one of the Gamma parameters

Smyth and Jorgensen p. 146 ASTIn Bulletin Vol 32(1) 2002

lambda is the Poisson rate, for given exposure time t, the rate will be lambda * t, so dispersion parameter depends on exposure time through t**(1 - p)    (same as what I had above)

**Comment by josef-pkt** [comment]
I'm just reading the Smyth and Jorgensen 2002 article and they use double exponential to allow dispersion to vary depending on explanatory variables, e.g. if the variance of the gamma claim size distribution varies with subgroups of the population or type of insurance policy separately from the poisson claim frequency.

**Comment by josef-pkt** [comment]
This reminds me now that we haven't checked the usage of exposure in Tweedie.

I finished reading Smyth and Jorgensen 2002 yesterday. It's a good paper that makes a clear distinction between claim cost only data, and the joint model for claim count and claim cost.

They use `w` everywhere to normalize for exposure, but I'm not sure what `w` really is and how this fit's in the code, i.e. exposure or var_weights. For the parameter estimates itself all weights and multiplicative exposure are the same (with appropriate scaling of endog). 

partially related: 
issues for weights and
issues for aggregation #2831 #2840

**Comment by thequackdaddy** [comment]
Josef,

`w_i` is the "units at risk". (First paragraph of section 2 of the paper.) I interpret that to be "exposure" more than `var_weights`. (I've been thinking of them as `freq_weights`). In other words, if I have a policy in force for 6 months, and it has 2 claims, the rate of claims per year (paper calls it frequency) is 4 (= 2 / 0.5). 

I'm thinking a full Tweedie approach would look (something) like this... 

1. Fit GLM `Tweedie` model with initial estimate of `p` (say `1.5`). (endog is dollars of loss / years  at risk)
2. Calculate new `p` using step 1 and the `estimate_tweedie_power` function. 
3. Refit 1, but use updated `p` from 2. 
4. Repeat between 2 and 3 until converged. 
5. use phi = lambda**(1-p) / (2 - p) to estimate `cov_params`. For lambda, use mean frequency of the entire dataset. `cov_params` will then reflect standard errors/z-scores that represent annual variation. 

So in other words, I'll have one phi that's used in the esimation of `p` and a separate phi that used only for calcualting `cov_params`. 

The only thing that (somewhat) concerns me here is that the paper makes it seem phi shoudl really be a vector... and different for each observation. Considering that some of my observations are 1 months of risk and others are 6 months of risk. I'm not convinced if having a vector of phi (as opposed to 2 scalars for these purposes) is of any value.

**Comment by josef-pkt** [comment]
(just briefly, comment about `w_i` will follow later)

step 5 doesn't work, steps 1-4 are good 

lambda also depends on the explanatory variables and varies by observation, AFAIU. 
The calculation go the other way, lambda for the compound Poisson is implied by our estimated parameters for mean and variance.

Now something is strange: because if lambda varies, then phi = lambda**(1-p) / (2 - p)  would also vary.
(We really need the functions to move between Tweedie and compound Poisson parameterization so it's easier to understand.)

in terms of expectation (mean)
mean(endog) = E(tweedie) = E(Poisson) * E(Gamma)

`cov_params` should be independent of any equivalent rescaling of the data.
But we can calculate functions of the estimated parameters that renormalize the parameters and calculate the standard errors for those.
e.g. for the "multiplicities" we want the exponential form of the parameters exp(params) with associated standard errors and confidence interval.

**Comment by thequackdaddy** [comment]
> Now something is strange: because if lambda varies, then phi = lambda**(1-p) / (2 - p) would also vary.

I've thought about that in the past as well. What I was thinking is that you *could* have a vector of phi instead of a scalar. I see that all the math would still work, with the exception of `cov_params` which is just `cov_params = scale * normalized_cov_params`. `normalized_cov_params` is a calculation for the WLS module = 

(Full disclosure, my only real interest in this whole issue is that `cov_params` is awfully high, when I'm confident it shouldn't be. I can a Poisson model on count data and have `cov_params` be quite reasonable. But with Tweedie, `cov_params` become really high. I presumed issue was that Pearson seems influenced when you split records. (I think this is an estimate of the hessian?)

```python
                self.normalized_cov_params = np.dot(self.pinv_wexog,
                                        np.transpose(self.pinv_wexog))
```
Granted, there's a lot of moving parts that go into `cov_params`... I could be looking in the wrong place. Its just that `scale` seems the most flexible concept we have, and the one I haven't gotten my mind wrapped fully around. (Especially scale when you have records with different `freq_weights` in the data...)

So I'm still thinking about this...

**Comment by thequackdaddy** [comment]
P.S... when I got involved with you, I really thought I'd be able to make a handful of fixes and be finished. It does seem that with stats, the deeper I go, the more I realize I don't know...  Hopefully there's a bottom to the ocean.

**Comment by josef-pkt** [comment]
There is no bottom to the ocean. But we might be at a level that is appropriate for what we want or need to do.

You could try `cov_type='HC0'` in fit, and see whether it's close to what you expect. It doesn't use the builtin `scale`.

I'm not quite sure about the interaction of exposure and frequency weights.
Do your frequency weights vary a lot?
You could check if you get cov_params closer to what you expect when you don't use freq_weights.

**Comment by josef-pkt** [comment]
Just a sanity check: integer division
Are you using python 2 or python 3? 
Did you check that your total freq_weights is equal to total nobs/exposure when you split the data?

**Comment by josef-pkt** [comment]
Also, if you can come up with a simple example that illustrates the effect, then it would be easier for me to look at it.

**Comment by josef-pkt** [comment]
Another check: Are `bse` (cov_params) tested against R?   
I thought so, but I might not have paid enough attention to which tests are against R and which are cross-checks.

**Comment by thequackdaddy** [comment]
- HC0... I've been meaning to say that I've had difficulty getting this to work on larger datasets. Haven't really played with it... so I've been using the default (`nonrobust`). I'll take a look later. 
- integer division - Using 3.5 atm. I was using 2.7 for something awhile back. 
- Well the data exists in "pre-split" format. So the splitting is a natural part of how the data is captured. In other words, I get records for expired policies. Some of those policies are 6 months, some are 1 week. (People might cancel 1 week into a policy term...)
- `bse` are tested against R. I pulled using this -> `sqrt(diag(s$cov.scaled))`

**Comment by josef-pkt** [comment]
You could also try the original data, total claim amounts without scaling, and use exposure for policy-years, e.g. exposure=0.5 for a half year policy. It only works with log link.

This is how we get the results normalized to exposure=1 in Poisson, when actual exposure fluctuates.
E.g. we can predict the mean at exposure=1 (which IIRC is the default)

(assuming exposure is correct in tweedie)

I can look at HC0, the main difficulty is that calculating the Hessian is slow right now.

**Comment by thequackdaddy** [comment]
[Here's a gist](https://gist.github.com/thequackdaddy/7a87783d8d165d319fd5f793488c3d99)

Assume that executions is actually executions per year. 

I just doubled the record. I assumed that all executions happen in the first half of the year. 

You get the same `params`, but everything else changes.

**Comment by thequackdaddy** [comment]
(For the record, you are not the first person I've asked this question too.... I asked a statistician at a consultant that sold a $150K GLM software to answer this question, and they couldn't...)

**Comment by josef-pkt** [comment]
If I use `endog2 = pd.concat((endog2 * 1, endog2 * 1))` then I get exactly the same result in both cases. (except that case two has nobs in summary instead of nobs, but the same df_resid.

So there is something specific to splitting it up this way. My **guess** is now that this changes the data, i.e. it's not supposed to be equivalent. You have now a lot more zeros in the dataset and in other cases you have much higher values. Even if it averages out for the mean, it will/might affect the variance/dispersion.

(BTW: Where do I need to send my invoice? :)

**Comment by josef-pkt** [comment]
Yes, I think that's it.
We need to estimate the dispersion coefficient because it's related to the Gamma variance. So we cannot fix it when we split. 

I tried the Poisson case: With the fixed scale=1, both cases produce the same result because of the additivity/homogeneity of the Poisson process.
However, in the second case the dispersion is much larger, so the dispersion would tell us that Poisson is not the right model because we have to much variation. If we scale the cov_params by the dispersion (pearson-scale), then the bse would also increase in the second model.

I think we don't have a function that calculates dispersion, so IIRC, for Poisson we get:
```
>>> (res1p.resid_pearson**2).sum() / res1p.df_resid
2.3977285254288887
>>> (res2p.model.freq_weights * res2p.resid_pearson**2).sum() / res2p.df_resid
11.458491444644904
```
which is an increase to almost 5 times the first dispersion.

**Comment by josef-pkt** [comment]
My guess is that if we observe the number of claims additionally, then we can impose time homogeneity on the arrival process, and identify/estimate the variance of claims separately. 
In that way we can impose this kind of additivity, but as in the Poisson case, the data would tell us that we don't have the right model. (e.g. too many zero observations for a Poisson arrival distribution)

As background: From what I think now, the behavior that you would like to get requires a very strong kind of additivity assumption.
If we put all accidents in the first half of the year and observe zero accidents in the second half, then you want it to have the same variance as when the accidents are distributed more uniformly over the year.
The mean prediction is still the same in both cases because the long term average doesn't change, but the large fluctuations are reflected in the time concentrated model.


maybe you want another property of the model to capture exposure.

**Comment by josef-pkt** [comment]
Zeroinflation is a good story for this.
Essentially you are introducing zero inflation into the model. If we were to estimate a ZeroInflatedPoisson model, then I would expect the zero inflation parameter to increase and at the same time the Poisson rate lambda so that it averages out to the same longterm mean.

Given that Tweedie and Poisson have an additive structure in the mean function, the mean function doesn't change, but the extra zeros are reflected in an increased dispersion, I guess which will imply a larger Gamma variance in the Tweedie compound Poisson case.

**Comment by josef-pkt** [comment]
typo in Smyth Jorgensen.

the "strange" think about phi above is that it is a typo in the article, on page 146

phi = lambda**(1-p) / (2 - p)

is missing the tau term and should be

phi = lambda**(1-p) tau**(2-p) / (2 - p)
tau is mean of a single Gamma claim

tau was referred to in the next paragraph in the paper, but there was nowhere the powr (2-p) in the equation.

If phi (dispersion coefficient) is assumed to be constant, then there is a fixed relationship between changes in lambda and tau such that
lambda**(1-p) tau**(2-p)  remains constant.

**Comment by josef-pkt** [comment]
I started to write a summary for the parameterization of Tweedie and compound Poisson
https://gist.github.com/josef-pkt/b8354b64959ebaf8a7b51c49cecdb54c
(math line breaks are messed up for some unknown reason.)

**Comment by thequackdaddy** [comment]
Yeah about that invoice... Haha... Happy to provide a donation to the josef-pkt charity fund for the acquisition of inebriating libations. Send me a bitcoin wallet/google pay wallet/paypal address. 

I don't know much about zero inflated poisson--though there's a [paper by Zhang](https://cran.r-project.org/web/packages/cplm/vignettes/cplm.pdf)--talking about a zero-inflated compound Poisson. I was always under the impression that this really dealt more with there being more density than expected at Zero because people don't want to file claims to keep a claims-free discount or because the report is naturally censored by a deductible. (Why report a $100 claim--or even a $501 claim--when I have a $500 deductible. Not going to get enough money to make it worth the effort.)  

I think you are right that I want a strong additivity assumption (and maybe the ability to test it somehow.) If I have a 1 year policy with a claim, I want to have the contribution to deviance for that policy to be the same as if it was 2 six month policies. (An insured can change a telephone number or something else minor that results in an endorsement of the policy. That creates a new record, splitting an annual record in 2 records that add up to a year.)

**Comment by josef-pkt** [comment]
I'm partway through coding the transformation function and a random sampler for compound poisson and tweedie. I hope it helps to look more closely at what's going on.

My intuition is also that there should be some additivity like this, maybe not strict, but we should at least roughly get the same standard errors. What is even stranger is that the standard errors bse are larger  in the split data than in the aggregated sample. If we have more detailed information, like per half year instead of per year observations, then the precision of our estimates should increase (unless the model is misspecified and we pick up more information that conflicts with the model, I think)

I think the way you split up the data and used frequency weights is not the right way to do this for inference. If I use exposure instead of frequency weights, or if I leave off the frequency weights, then the bse is only a little bit larger in the split model than in the original model.

---
On the other hand, additivity might not be possible without including the count, number of claims in an observation. (maybe unless there is only a very small probability to have more than one claim per period, so it's essentially just zero or one claim).
When there is a large payment in one period, then tweedie has to allocate the total payment between number of claims and size of a single claim. It could be that tweedie is a good, approximately correct model for some aggregation level but a bad model for others.
For example (I didn't try to read anything and don't know any details, but it's mentioned in Jorgensen articles). roughly: some distributions converge to a Tweedie distribution similar to convergence to the normal distribution by the law of large numbers. So Tweedie could be good approximation for aggregated samples but could be a inappropriate, misspecified model for disaggregated data.

When we work with actual data, it can be difficult to guess what are the effects of a model (given that we don't know all the statistical theory) and what are the effects of using a misspecified model.

**Comment by josef-pkt** [comment]
estimating a simulated tweedie with nobs = 10000

just a simple two group example
true values p = 1.5, scale (phi) = 2.0
mean in two groups 5 and 10
```
In [104]: res1.model.estimate_tweedie_power(res1.fittedvalues), res1.scale
(1.4794568197139488, 1.9920646777181317)

In [106]: np.exp(res1.params.cumsum())
array([  4.89525573,  10.03595213])
```

looks good

**Comment by josef-pkt** [comment]
to continue on this: a Monte Carlo with 1000 replication, nobs=10000

```
time 46.37501406669617
[ 1.49822651  1.99951565]   # MC mean of power, dispersion, true is 1.5, 2.0
[ 0.00206508  0.00099   ]   # MC variance
```

looks very good in large sample and true model is Tweedie compound Poisson, no bias here

**Comment by josef-pkt** [comment]
I updated the gist https://gist.github.com/josef-pkt/b8354b64959ebaf8a7b51c49cecdb54c

Add the end I did a similar experiment by summing each consecutive two observations (which have the same exog). This means that initially I have half year data that I combine into one year data. (exposure is per half year) (or we could take initially one year and then we combine two years)

Using exposure, I get exactly (at a few decimals) the same numbers in both cases.
With frequency weights the bse differ a bit (assuming each combined observation has frequency 2 with half the endog in the last case in notebook).
Without frequency weights (freq_weights=1), the bse is the same with the 2period aggregate as in the separate period case, but the estimated means double, because we have no normalized to 2 period exposure.

So my guess is that frequency weights don't do the appropriate thing for this case. The correct interpretation requires `exposure`, and endog as the total claims for the exposure time.
Another guess is that we might need var_weights if the endog is a mean over some observations.

This is with simulated tweedie in a large sample, so it rules out misspecification of the underlying model.

**Comment by thequackdaddy** [comment]
Oh my. Thank you.

There's a very simplistic analogy to this behavior [here](http://www.theanalysisfactor.com/the-exposure-variable-in-poission-regression-models/).

For Tweedie, my endogenous really needs to be the loss and an offset for exposure is the used.

This is a super helpful notebook.

**Comment by thequackdaddy** [comment]
(That $150k software really doesn't explain things as well as you do.)

**Comment by thequackdaddy** [comment]
@josef-pkt I have another problem I want to talk through. So bear with me. 

I agree--in general--that `exposure` is the correct thing to use here. I think `var_weights`  is incorrect but I'm not sure that the assumption because I'm not sure so certain that dispersion--`(endog - mu) ** 2`--scales linearly time. In fact, I would think this is not the case. (Let me know if my understanding of dispersion is flawed.)

Let's go back to [this paper](https://hal.archives-ouvertes.fr/hal-00746104/document). 

Recall that `S_i ** 2` is:

![image](https://cloud.githubusercontent.com/assets/8186432/16359624/8b5a2f84-3b02-11e6-872e-f4e93f9398e4.png)

(Note to self, the paper is from a bio-statistics perspective. We essentially have "observations"--which the paper calls sites--and each observation has 1 replicate. I presume that just dropping the `1 / (n_i - 1)` piece is harmless.)

The author presents this formula... 

![image](https://cloud.githubusercontent.com/assets/8186432/16359695/70fd0f24-3b04-11e6-9a1e-7a2fc15f3522.png)

And using a log\gamma GLM, we should be able to solve for `a` and `b`. 

However, when doing this on my 3.5 million record dataset, it didn't work. I read this section...

![image](https://cloud.githubusercontent.com/assets/8186432/16359701/d1758dd6-3b04-11e6-9afa-63f916401a23.png)

Should I be weighting this somehow, perhaps? Maybe based on `Exposure`? I played with some different ideas and think this is probably as close to something reasonable as I could find...

```python3
y = (model.endog / np.exp(model.exposure) - res.mu / np.exp(model.exposure)) ** 2
x = np.column_stack((np.ones(len(res.mu)),
                     np.log(res.mu / np.exp(model.exposure))))

model2 = sm.GLM(y, x, freq_weights=np.exp(model.exposure),
                family=sm.families.Gamma(sm.families.links.log))
res2 = model2.fit(maxiter=5)
```
Which gives... 
```python3
In [103]: res2.summary()
Out[103]: 
<class 'statsmodels.iolib.summary.Summary'>

                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      y   No. Observations:              3569086
Model:                            GLM   Df Residuals:                   456160
Model Family:                   Gamma   Df Model:                            1
Link Function:                    log   Scale:                   72062.4992586
Method:                          IRLS   Log-Likelihood:            -9.5363e+06
Date:                Sat, 25 Jun 2016   Deviance:                   5.6396e+06
Time:                        18:54:39   Pearson chi2:                 3.29e+10
No. Iterations:                     5                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.4552      4.448      2.351      0.019       1.737      19.173
x1             1.1295      0.919      1.229      0.219      -0.671       2.930
==============================================================================
```

`np.exp(10.4552) == 34724.473308301465` vs the pearson x2 estimate of 6147.95962013. (So not good.)

However, `p == 1.1295` is believable to me...

On a side note, I now think that that our shortcut doesn't work well when you have different exposures. I'm fairly certain that we might need to adjust this to account for exposures better. I just don't know where/how to do it... I appreciate any ideas. Recall that we plug 3.8 into 3.9, and then use the `brentq` optimizer to set this equal to 0.

![image](https://cloud.githubusercontent.com/assets/8186432/16359751/4cf83f60-3b07-11e6-9aaf-372f974ced19.png)

Thanks.

**Comment by josef-pkt** [comment]
I'll try to look at it tonight (Central European Time), I don't remember the details now.

Yes, I think the pearson estimate needs to take account of exposure as in (3.9) in your comment. This needs to be fixed (for 0.8) if it doesn't do it correctly.

**Comment by josef-pkt** [comment]
in estimate_tweedie_power, we could try to replace endog and mu by the exposure corrected values

```
endog = self.endog / exposure
mu = mu / exposure
```

where exposure is the correct exponentiated exposure (internally still stored as log exposure, which I think hasn't changed yet). and replace self.endog by endog in all subsequent expressions.

However, I'm not able to figure out right now what the interpretation is. Even without sun, beach, drinks and three kids it will take me some time to remember the different aggregation arguments :)
Essentially, I don't remember how dispersion should scale with exposure given that mu also increases, but I think correcting for exposure in estimate_tweedie_power is most likely correct.

BTW: we should still add a `method="gamma-glm"` to estimate_tweedie_power, but I got busy with other things before I managed to produce any code for double exponential family.

**Comment by thequackdaddy** [comment]
For gamma-glm, I did that in a branch that I'm working on. I'll make these changes and send to you.

I *think* if we do this, we ought weight the calculations based on exposure. The way I interpret it is that each row in my design matrix is equivalent to a "site" and and the replicates (n_i in the paper) are the amount of exposure. We should weight using number of exposures too. I think that makes sense to me.

I'll play with the this some more. 

Enjoy the drinks. I really should buy you one (or 100)...

**Comment by thequackdaddy** [comment]
Ok, I've managed to confuse myself. 

From Taylor's Power Law for Dependence of Variance on Mean in Animal Populations, Perry (1981) available on jstore. 

<img width="648" alt="screen shot 2016-06-26 at 10 13 02 pm" src="https://cloud.githubusercontent.com/assets/8186432/16367664/6dc8619a-3beb-11e6-8cd1-5f78f92950f0.png">

Note that j indexes the sites, and i indexes the time (which I presume is basically the same thing as replicates.)

From Bias-corrected Pearson estimating functions for Taylor’s power law applied to benthic microfauna data (Jorgensen, 2011), available at link above...

<img width="744" alt="screen shot 2016-06-26 at 10 18 57 pm" src="https://cloud.githubusercontent.com/assets/8186432/16367732/0b399200-3bec-11e6-9aae-4acbc206be1a.png">

<img width="779" alt="screen shot 2016-06-26 at 10 20 03 pm" src="https://cloud.githubusercontent.com/assets/8186432/16367760/4245d5ba-3bec-11e6-9508-d3e5d1773831.png">

So notice that in the Perry paper, the sigma summation for the variance sums up all the sites (indexed by j) at time i. 

In the Jorgensen paper, the sigma summation for the variance sums up all the replicates (indexed by j) at site i. 

I'm not sure if this is an error, or just means that you can do the calculation either way. 

I wonder if I could use Taylor's power formula and back into 'b' but substituting Var(Yij) for S_i^2 and some kind of average on endog. Alternatively, my insurance data is annual. I could group the data by year (or year and month if I want more samples), calculate S_i^2 and mu_i within each each year, and then use the regression formula outlined to estimate scale/shape...

Let me sleep on this... I might be understanding how this works better... 

It could also be the porter I just drank talking.

**Comment by thequackdaddy** [comment]
After contemplating this over again, I've come to a different conclusion. (No surprise..)

I re-read some articles by Glenn Meyers and poked around and found out that a lot of the old-school Tweedie folks in actuarial circles simply assume that the distribution is Poisson/Gamma, with no zero-inflation, and then using some simple algebra, realize that you only need to calculate the coefficient of variation of the losses to back into an estimate of alpha parameter for the gamma distribution, then you back into an estimate of p. I get an answer of 1.6 for Tweedie when I do do that on my real dataset. (Approx 3.5 million records representing about 450,000 exposures.)

From [this newsletter by Glenn Meyers](http://www.casact.org/newsletter/pdfUpload/ar/Brainstorms_corrected_from_AR_May2009.pdf)


```python
In [115]: z = np.array(data['PDIncurredLoss'])

In [116]: n = np.array(data['PDClaimCount'])

In [117]: y = np.where(n > 0, z / n, 0)

In [122]: mu_y = (y * n).sum() / n.sum()

In [123]: mu_y
Out[123]: 2683.7337468276892

In [124]: var_y = ((y ** 2 * n).sum() / n.sum()) - (mu_y **2)

In [125]: var_y
Out[125]: 10792740.967788128

In [126]: cv2 = var_y / (mu_y ** 2)

In [127]: cv2
Out[127]: 1.4984867228406344

In [128]: alpha = 1 / cv2

In [129]: (alpha + 2) / (alpha + 1)
Out[129]: 1.5997577290052363
```
I'm thinking my acceptance of the ~1.15 p value earlier was grossly mistaken.

The implementation of `'brentq'` I committed yesterday is now feeling a little better. That is calculating a `1.645`, which makes sense. Intuitively from Meyer's formulas, you expect a number around 1.5 if you the gamma distribution coefficient of variation is around 1. If its a little greater (which it is based on my math above), you want something above 1.5. A little interesting thing about what I implemented is that I use the unadjusted values to calculate the Pearson scale parameter, then the adjusted `endog` and `mu` in the part where the scipy optimizer runs. I also multiply that part by weight, (which I've tenatively defined as the product of `freq_weights` and `exposure`. 

I'm leaning towards thinking that the implementations of the `'gamma-glm'` and `'ols'`, need work though. I'm hoping to make a few more datasets to provide some clarity to myself on this. 

Side note, many of papers I've read use (what I think) is `var_weight` instead of `exposure`, and thus endog is claim payments per unit of exposure, while I'm just modelling claim payments. Intuitively, I've come to like `exposure` as a better concept (I don't like the assumption that variance scales linearly with time, if I understand correctly). 

I have some other projects to work on, so I might pause this for a minute to think this through.

**Comment by thequackdaddy** [comment]
Note to self:

Did some more book learning and a few tests using real data.

* My bentq method I still don't have an intuitive feel for how it really works. It worked on the largest dataset that I had, but failed miserably on the second largest. The second largest is a less homogenous coverage where the claims can be really small, or really big.. a few hundred dollars to $100k or more. These claims are typically rare, however. These claims are likely to end up in litigation, so they are expensive. Wondering if a small, homogenous claim work work, but a rare, volatile form of claims wouldnt't work.
* I found a paper with another method for estimating `p` and `scale`. The `scale` essentially is the deviance scale and the `p` is derived from it, as far as I can tell. There's a formula that you have to maximize. I presume the spicy optimizers could be helpful. 

[Wuthrich](http://www.actuaries.org/LIBRARY/ASTIN/vol33no2/331.pdf)

**Comment by josef-pkt** [comment]
@thequackdaddy  Besides fixing exposure or whatever might not be correct:
I think in your case you need more diagnostics, goodness-of-fit measures and comparison with alternative models to see whether Tweedie is an appropriate distributional assumption for your data.

E.g. It could be that the gamma has not thick enough tails to model your very large values without distorting the distribution like outliers.
If you know the number of claims,, you can also still estimate the poisson and gamma parameters separately and see whether they agree approximately with the implied Tweedie paramaters.

**Comment by thequackdaddy** [comment]
I'm rather comfortable with Tweedie. 

Here's [a gist](https://gist.github.com/thequackdaddy/08f3006d61350752bca963f32f31481b) of the data. There are 2 claim types, bodily injury (called BI) and property damage (called PD).

Bodily injury is much more likely to have litigation and large claims, you can see that the claim buckets are quite wide. 

Property damage claims are a little tighter. 

I'll work on goodness of fit tests later this weekend, in all likelihood.
